File: 70-ClimbingStairs/70-ClimbingStairs.py
# Last updated: 7/2/2025, 11:54:29 PM
class Solution:
    def climbStairs(self, n: int) -> int:
        '''no_ways = 0
        dp = [-1]*(n+1)
        return self.help(n,dp)
        #return no_ways'''
        dp = [0] * (n+1)
        dp[0],dp[1] = 1,1

        for i in range(2,n+1):
            dp[i] = dp[i-1]+dp[i-2]

        return dp[n]
    def help(self ,n:int, dp:List[int]) -> int:
        if(dp[n] != -1):
            return dp[n]
        if(n <= 1):
            return 1
        dp[n] = self.help(n-1, dp) + self.help(n-2, dp)
        return dp[n]
        

File: 70-ClimbingStairs/70-ClimbingStairs.cpp
// Last updated: 7/2/2025, 11:54:25 PM
class Solution {
public:
    int climbStairs(int n) {
        // if(n < 0  ) return 0;
        // if( n == 0) return 1;

        // int one = climbStairs(n-1);
        // int two = 0;
        // if( n > 1) two = climbStairs(n-2);

        // return one + two;

        vector<int> dp(n+1); // if it is 1 based indexing

        if(n <= 2 ) return n;
        dp[0] = 0;
        dp[1] = 1;
        dp[2] = 2;

        

        for(int i=3;i<=n;i++){
            dp[i] = dp[i-1] + dp[i-2];
        }

        return dp[n];
    }

    void help(int n, int &ways)
    {
        if(n < 0  ) return;
        if( n == 0){
            ways++;
            return;
        }

        help(n-1, ways);
        //int two = 0;
        if( n > 1) help(n-2, ways);

    }
};